Proper export of the libraries for external use of kml

johan-andruejol opened the pull request:
The find_library failed on my windows machine (kml library not found). Using the target is better for exporting target and fixes the issue.

johan-andruejol committed:
Proper export of the libraries for external use of kml

mwoehlke-kitware commented on CMakeLists.txt:
```diff
@@ -275,11 +275,14 @@ IF(WIN32)
+  LIST(APPEND KML_TARGETS expat)
 ELSE (NOT LIBKML_USE_EXTERNAL_KML)
   TARGET_LINK_LIBRARIES(kml ${EXPAT_LIBRARIES})
+  LIST(APPEND KML_TARGETS ${EXPAT_LIBRARIES})
```
Something's not right here: when I try to configure this, I get:

```
CMake Error at CMakeLists.txt:384 (export):
  export given target "expat" which is not built by this project.
```

I'm pretty sure expat should not be an exported target unless the embedded version is being built; when using an external expat, it won't be a target.

mwoehlke-kitware commented on KMLConfig.cmake.in:
```diff
@@ -4,4 +4,6 @@ set(KML_INCLUDE_DIRS
 
 set(KML_LIBRARY_DIR  "${libkml_BINARY_DIR}")
-find_library(KML_LIBRARY kml PATHS "${libkml_BINARY_DIR}" NO_DEFAULT_PATH)
+if(EXISTS "@KML_TARGETS_FILE@" AND NOT TARGET kml)
```
This is wrong; `KMLConfig.cmake` should look for `KMLTargets.cmake` in the same directory as itself (). It should _not_ embed the build directory, as this will be wrong for installs, especially relocatable installs.

Depending on what version of CMake you're willing to require, instead use either `CMAKE_CURRENT_LIST_DIR` or:

```
get_filename_component(_SELF_DIR "${CMAKE_CURRENT_LIST_FILE}" PATH)
```

johan-andruejol committed:
Do not export expat if it's external

johan-andruejol committed:
Fix the KMLTargets call in KMLConfig*

The KMLTargets should always be alongside the KMLConfig (both build and
install). We can use CMAKE_CURRENT_LIST_DIR since CMake required version
is 2.8.4

johan-andruejol commented:
Thanks for the comments !
I pushed a newer version that should address them :)

mwoehlke-kitware commented on KMLConfig.cmake.in:
```diff
@@ -4,4 +4,6 @@ set(KML_INCLUDE_DIRS
   "${Boost_INCLUDE_DIR}")
 
 set(KML_LIBRARY_DIR  "${libkml_BINARY_DIR}")
```
Since we are now exporting targets, I think this is superfluous? If yes, `KMLConfig.cmake.in` and `KMLConfig-install.cmake.in` can be collapsed into one file, which would be nice :smile:.

johan-andruejol replied:
Yes, Good point !

johan-andruejol committed:
Collaspe KML configuration into one file only

johan-andruejol committed:
Fix LIBKML_USE_EXTERNAL_EXPAT typo

This typo prevented the correct behavior for the linking/exporting of the
expat libraries, causing expat to always be considered as build by KML.

johan-andruejol commented:
@mwoehlke-kitware @aashish24 
What is the status of this ? Can it be merged ?

mwoehlke-kitware commented:
Seems okay to me; if @aashish24 doesn't comment/object within a few days, I think we can merge this. (Ping me again if I forget :grin:.)

johan-andruejol commented:
@mwoehlke-kitware Ping :)

johan-andruejol commented:
Awesome ! Thanks !
